### ISSUE TITLE
deoptimization warnings

### ISSUE THREAD

**Issue by mscdex** [jshint/jshint#2054]
I'm wondering if it's possible to add some optimization warnings (especially for node.js/v8). For example, checking if a function both uses `arguments` and reassigns function parameter variable values:

``` javascript
function foo(bar, baz) {
  bar = 'quux';
  if (arguments.length < 2) {
    // ...
  }
}
```

v8 will disable optimizations for such functions and it would be nice to be able to easily check for these kinds of situations when linting with jshint.

**Comment by impinball** [comment]
Ran the following script in the latest V8 build (through d8):

```js
function testOpt(fn) {
  var list = fn();
  %OptimizeFunctionOnNextCall(fn);
  list = fn();
  switch (%GetOptimizationStatus(fn)) {
    case 1: return fn.name + " is optimized";
    case 2: return fn.name + " is not optimized";
    case 3: return fn.name + " is always optimized";
    case 4: return fn.name + " is never optimized";
    case 6: return fn.name + " is maybe deoptimized";
  }
}

function foo() {
  return [];
}

function bar() {
  return foo.call(null, arguments);
}

print(testOpt(foo));
print(testOpt(bar));
```

Here's my output with just `--allow-natives-syntax`:

```
foo is optimized
bar is not optimized
```

So, clearly there is a deopt going on. So, I ran it with `--allow-natives-syntax --trace_opt --trace_deopt`:

```
[disabled optimization for 0x3ee6f545bd51 <SharedFunctionInfo InstantiateFunction>, reason: TryCatchStatement]
[disabled optimization for 0x3ee6f5467439 <SharedFunctionInfo ConfigureTemplateInstance>, reason: TryFinallyStatement]
[disabled optimization for 0x3ee6f5453d89 <SharedFunctionInfo DefaultNumber>, reason: Call to a JavaScript runtime function]
[compiling method 0x22f8db54a191 <JS Function foo (SharedFunctionInfo 0x22f8db549eb9)> using Crankshaft]
[optimizing 0x22f8db54a191 <JS Function foo (SharedFunctionInfo 0x22f8db549eb9)> - took 0.024, 0.099, 0.036 ms]
foo is optimized
[compiling method 0x22f8db54a1d9 <JS Function bar (SharedFunctionInfo 0x22f8db549f49)> using Crankshaft]
[aborted optimizing 0x22f8db54a1d9 <JS Function bar (SharedFunctionInfo 0x22f8db549f49)> because: Bad value context for arguments value]
[disabled optimization for 0x22f8db549f49 <SharedFunctionInfo bar>, reason: Bad value context for arguments value]
bar is not optimized
```

I think the output confirms it. So, even in the current tip of tree for V8, this is unoptimizable. IIRC, it's an automatic bail because of cases such as remotely modifying arguments. Prepending a `'use strict'` directive had no effect.

**Comment by caitp** [comment]
> I think the output confirms it. So, even in the current tip of tree for V8, this is unoptimizable. IIRC, it's an automatic bail because of cases such as remotely modifying arguments. Prepending a 'use strict' directive had no effect.

It's a bit more complicated than that --- for details you can read hydrogen.cc. But even beyond that, it's even more complicated for other reasons: the intermediate graph can be reduced and improved further to allow previously bailed optimizations at a later time. This is also crankshaft and v8 specific, so I don't feel great about telling people "YOU CAN NEVER DO X, Y or Z" because A) it can change, and B) it affects a specific engine at a specific time under a specific configuration

**Comment by jugglinmike** [comment]
You can use static analysis to make suggestions about inefficient code, and JSHint is a static analysis tool, but this problem space has a lot of unique concerns.

Thinking from a high-level API perspective: a new option like `optimizable: true|false` would fit aesthetically, but it wouldn't really address all the concerns. As @caitp mentions, things change over time and between platforms. To do this problem justice, it seems like you'd need a lot more configuration (e.g. the specific platforms and versions that you are targeting).

This sounds like an interesting problem, but trying to solve it within JSHint itself doesn't seem quite right to me. Anton decided (correctly in my opinion) that relegating code style checks to a separate project will allow JSHint to be more focused and stronger overall. I think the same could be said for code optimization.

At the same time, [the options documentation](http://jshint.com/docs/options/) cites optimization as the secondary rationale for certain features, so maybe my interpretation is actually contrary to the goals of the project.

**Comment by mscdex** [comment]
Well there is already a `node` config option that would imply v8, unless node should change JS engines in the future... Whether or not the code optimization checks are implemented in a separate project, it would be nice to have one tool to show both the current jshint checks and the code optimization checks.

**Comment by impinball** [comment]
Yes, it would be nice, since Node exclusively uses V8, and a lot of Node
developers will tend to care about this.

**Comment by impinball** [comment]
@jugglinmike 

V8 does have those kinds of primitive blacklists. Here's the [exact line](https://chromium.googlesource.com/v8/v8.git/+/fb2643c858c96a783fced0c730a6ff89ea60eda5/src/hydrogen.cc#4044) this is bailing from.

**Comment by caitp** [comment]
It is worth looking at it from a different perspective. So, ignoring that crankshaft will deopt in these cases, for the time being, consider this: 1) The cases where `arguments` can be used without deopt is a bit complicated. In HEAD, you can't assign to `arguments` or any value of `arguments`, you can't push `arguments`, you __can__ use `arguments` in `Function.apply()` if each parameter is stack-allocated, bunch of other subtle rules... We don't really have a good way of tracking all of this.

It's probably more appropriate to use deopt tracing during a test run in order to find where deopts are happening, and why, and with your specific version of V8.

But, a more important point is this: deoptimization often doesn't matter. It would be better to encourage people to profile their code and find what areas of their code are "hot", before complaining every time they do something that makes their code easier to read, but might result in a deopt bailout (such as using try/catch in a function vs putting try/catch in a separate function, as is done in Bluebird, with likely minimal benefit)

I just don't think static analysis is the right way to solve this problem, and will likely encourage people to write worse code than they otherwise would

**Comment by mscdex** [comment]
For what it's worth, I can vouch for the try-catch-in-a-separate-function optimization, I've seen significant gains from it.

Deoptimization may not matter in some edge cases, but most people don't profile their code (or at least regularly) and so are unaware that their code could be executing faster with some simple changes. I'm not saying that the static analysis has to be extremely complicated and get optimization checks exactly right, the warnings are just that: warnings (about possible slow cases or deopts). Anyway, I just thought it might be worth adding at some point in the future.

**Comment by impinball** [comment]
And I will also note this: I would think, assuming it is implemented as an option, it being disabled by default, and not implied by `enforceall`. Most of its use cases are in the Node (and eventually IO.js) world.

**Comment by impinball** [comment]
Okay. I do also agree that this isn't something that static analysis should
really get into. It's just a good excuse to actually profile your code.
